File: cpp/SingleNumber.cpp
# include <iostream>
# include <vector>
# include <set>
# include <numeric>

using namespace std;

class SingleNumber
{

public:
    int singleNumber1(vector<int>& nums){
        // O(n) time, O(n) space
        set<int> s(nums.begin(), nums.end());
        return accumulate(s.begin(), s.end(), 0) * 2 - accumulate(nums.begin(), nums.end(), 0);
    }
    int singleNumber2(vector<int>& nums){
        // O(n) time, O(1) space
        int s = 0;
        for (auto x: nums)
            s ^= x;
        return s;
    }
};

int main(int argc, char const *argv[])
{
    vector<int> nums = { 25, 1, 1, 3, 3, 5, 5};

    SingleNumber sn;

    cout << "Using set+math: " << sn.singleNumber1(nums) << "\n";
    cout << "Using bits: " << sn.singleNumber2(nums);

    return 0;
}

File: py/SingleNumber.py
def single_number1(nums: list[int]):
    # using sets, O(n) time, O(n) space
    return 2 * sum(set(nums)) - sum(nums)

def single_number2(nums: list[int]):
    # using bits, O(n) time, O(1) space
    s = 0
    for num in nums:
        s ^= num
    return s

if __name__ == '__main__':
    nums = [25,1,2,1,2]
    num = single_number1(nums)
    print('Using sets+math:', num)
    print('Using bits:', num)